Automattic/woocommerce-payments
# Reset duplicate notifications from new plugins

Closes https://github.com/Automattic/woocommerce-payments/issues/8809

#### Changes proposed in this Pull Request
This PR ensures that a previously dismissed duplicate notice on the Settings page will reappear if a new plugin triggers duplicates again. This is accomplished by enhancing the dismissed notification object stored in `wp_options`. Previously, only the payment method was stored. With this PR, we also store the list of gateway IDs of plugins causing duplicates. This enhancement allows us to handle scenarios where new plugins cause duplicate notices.









#### Testing instructions
> 💡 In order to test multiple plugins enabling the same payment method, you can either follow the instructions in https://github.com/Automattic/woocommerce-payments/pull/8513, which mention installing/configuring plugins like Mollie, Stripe and Klarna, or add the following line of code in [this place](https://github.com/Automattic/woocommerce-payments/pull/8919/files#diff-6022afe6b12c5f150ff6dd0e1fe8f44fdc67bfb7bd39aa64dccc892b74c76418R63): **$this->gateways_qualified_by_duplicates_detector['card'][] = 'plugin_gateway_id';** (more details in the testing steps)
In the testing instructions for this PR, I will follow the second approach to save a bit of time, but feel free to proceed with installing & configuring plugins which is described in the PR mentioned above.

```
Quick reminder on how the inline notice looks like:
```
<p align="center">
<img width="501" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/22319444/9fd098c6-4fcc-489b-8aab-67743fc813c5">
</p>

* Before [this line](https://github.com/Automattic/woocommerce-payments/pull/8919/files#diff-6022afe6b12c5f150ff6dd0e1fe8f44fdc67bfb7bd39aa64dccc892b74c76418R63), add the following code snippet to simulate Stripe plugin enabling card and Apple Pay/Google Pay 
```php
$this->gateways_qualified_by_duplicates_detector['card'][] = 'stripe';
$this->gateways_qualified_by_duplicates_detector['apple_pay_google_pay'][] = 'stripe_apple_pay_google_pay';
```

* Navigate to Payments settings and you will see the inline notification about duplicates for both payment methods
* Dismiss for both payment methods and confirm they _both_ don't appear upon page refresh
* Simulate one more enablement of credit card w/o removing the previously added lines
```php
$this->gateways_qualified_by_duplicates_detector['card'][] = 'mollie_cc';
```
* **Refresh Settings page and confirm you see the duplicates notification again** - this is the main goal of this PR
* Dismiss the notification for credit card, refresh the page and confirm it's not reappearing
* Remove two last lines of code you previously added in `class-duplicates-detection-service.php` by keeping only `card` from `stripe` now, refresh the Settings page
    * upon page refresh, there should be no notification displayed again
    * this is important because we want to ensure that if a user dismisses notifications from multiple plugins in sequence, and then disables one of those plugins, the dismissed notifications from the other plugins remain hidden. For example, if notifications from the second and third plugins are dismissed, turning off the third plugin should not cause the second plugin's notification to reappear, as it was dismissed before

> **_Optional_**: after going through all the testing instructions, the `wp_options` table in your local database should contain the `wcpay_duplicate_payment_method_notices_dismissed` option with the following values. It's not a must-have to check this (feel free to do so if you wish), but rather, it provides additional information on what happened in the background when we dismissed notifications for multiple plugins.

```php
Array
(
    [card] => Array
        (
            [0] => stripe
            [1] => mollie_cc
        )

    [apple_pay_google_pay] => Array
        (
            [0] => stripe_apple_pay_google_pay
        )

)
```

-------------------


**Post merge**
